### DISCUSSION TITLE
v2.0.0-rc1

### DISCUSSION THREAD

**Discussion by mosajjal** [mosajjal/sniproxy#49] [Announcements]
## What's Changed
* Bump docker/metadata-action from 4.3.0 to 4.4.0 by @dependabot in https://github.com/mosajjal/sniproxy/pull/37
* V2 by @mosajjal in https://github.com/mosajjal/sniproxy/pull/48

**Full Changelog**: https://github.com/mosajjal/sniproxy/compare/v1.0.1-beta...v2.0.0-rc1

**IMPORTANT**: sniproxy v2 is a complete overhaul over v1. no more command line option or JSON configs, everything is configured through a YAML file (and some CSV files). refer to the default config for all the configuration documents

<hr /><em>This discussion was created from the release <a href='https://github.com/mosajjal/sniproxy/releases/tag/v2.0.0-rc1'>v2.0.0-rc1</a>.</em>

**Comment by holygrolli** [comment]
as udp dns and https sni proxying is working fine now I tried to further improve my setup by setting up DoH. But this does not work when I am using a socks5 server.

```
2023-05-13T22:24:40Z DBG app/https.go:49 > overriding destination IP <nil> with 127.0.0.1:39085 service=https
2023-05-13T22:24:40Z INF app/https.go:65 > establishing connection to 127.0.0.1:39085 from clientip:51045 with SNI doh.myhost.example service=https
2023-05-13T22:24:40Z INF app/https.go:82 > could not connect to target with error: socks connect tcp socks5ip:1080->127.0.0.1:39085: unknown error connection not allowed by ruleset service=https
2023-05-13T22:24:40Z INF app/https.go:82 > could not connect to target with error: socks connect tcp socks5ip:1080->127.0.0.1:39085: unknown error connection not allowed by ruleset service=https
```
So as I understand the override rule the `Decision` is set to `override` and the `https` package is then trying to call through the socks5 like with every outgoing request. On a quick search I didn't find a hint on how to teach ther `Dialer.Socks5` to not use the proxy for any IPs. I tried to set `NO_PROXY=127.0.0.1` but this did not work.

**Comment by mosajjal** [comment]
interesting issue. technically, it's possible to exclude localhost from SOCKS5 upstream and always go direct. shouldn't be a big change. it'd be good if we could add unit tests for these too actually.

**Reply by mosajjal** [reply]
e4701887036b8af98aafa26e8768516d52382748 potentially solves the issue

**Reply by holygrolli** [reply]
ok it doesn't use the socks5 any more but it still acts unexpected:
```
2023-05-13T22:24:40Z DBG app/https.go:49 > overriding destination IP <nil> with 127.0.0.1:39085 service=https
2023-05-13T22:24:40Z INF app/https.go:65 > establishing connection to 127.0.0.1:39085 from clientip:51045 with SNI doh.myhost.example service=https
2023-05-13T22:24:40Z INF app/https.go:49 > overriding destination IP <nil> with 127.0.0.1:39085 service=https
```
does a DoH request on my `doh_sni` need to use the `tcpproxy`? probably because dohserver is listening separately on a different port?

and my DoH request from client keeps unanswered:
```
* Connection #1 to host doh.myhost.example left intact
* a DoH request is completed, 1 to go
< HTTP/2 404
< content-type: text/plain; charset=utf-8
< x-content-type-options: nosniff
< content-length: 19
< date: Sun, 14 May 2023 06:27:16 GMT
<
* Connection #2 to host doh.myhost.example left intact
* a DoH request is completed, 0 to go
* DoH: Bad ID type A for host.com
* DoH: Bad ID type AAAA for host.com
* Closing connection 0
curl: (6) Couldn't resolve host name
```

**Reply by mosajjal** [reply]
the DoH listener in `sniproxy` up-streams to the UDP listener on localhost, and `sniproxy` doesn't support UDP over SOCKS5(yet). I just pushed a couple of commits for an experimental SOCKS UDP client. 

```
Client ---direct---> sniproxy:443 ---direct---> sniproxy:randomport(doh) ---direct----> sniproxy:53udp ----socks5----> internet 
```

**Reply by holygrolli** [reply]
I am confused about why is socks5 involved at all? My expectation would be something like this
```
Client ---direct---> sniproxy:443 ---direct---> sniproxy:randomport(doh) ---direct----> sniproxy:53udp
```
After reaching the local sniproxy dns on port 53 the DoH should answer either with OriginIP or SniproxyIP. 
And from my log above I cannot see anything reaching sniproxy on port 53 at all

**Reply by holygrolli** [reply]
sorry for posting here several times but I had time to dig through the code and found out the `doh_sni` host is not enough for the DoH client, we have to add [`/dns-query`](https://github.com/mosajjal/sniproxy/blob/ee70cab915102d221410e9aecb13e51e66c23d26/dohserver/server.go#L69) to the host. now cURL can resolve the IP but it is trying to do a HTTP connect (is this DoH spec or cURL related, I don't know) which itself is being rejected by sniproxy because of https://github.com/mosajjal/sniproxy/blob/ee70cab915102d221410e9aecb13e51e66c23d26/httpproxy.go#L70 and `Decision` being `ProxyIP`. Is there a reason?

I will test further

**Reply by mosajjal** [reply]
> I am confused about why is socks5 involved at all? My expectation would be something like this
> 
> ```
> Client ---direct---> sniproxy:443 ---direct---> sniproxy:randomport(doh) ---direct----> sniproxy:53udp
> ```
> 
> After reaching the local sniproxy dns on port 53 the DoH should answer either with OriginIP or SniproxyIP. And from my log above I cannot see anything reaching sniproxy on port 53 at all

if you're answering the DNS query with SniproxyIP, you're right. but for getting the originIP, `sniproxy` needs to ask the upstream DNS server and get the real data.

**Reply by mosajjal** [reply]
> sorry for posting here several times but I had time to dig through the code and found out the `doh_sni` host is not enough for the DoH client, we have to add [`/dns-query`](https://github.com/mosajjal/sniproxy/blob/ee70cab915102d221410e9aecb13e51e66c23d26/dohserver/server.go#L69) to the host. now cURL can resolve the IP but it is trying to do a HTTP connect (is this DoH spec or cURL related, I don't know) which itself is being rejected by sniproxy because of
> 
> https://github.com/mosajjal/sniproxy/blob/ee70cab915102d221410e9aecb13e51e66c23d26/httpproxy.go#L70
> 
> and `Decision` being `ProxyIP`. Is there a reason?
> I will test further

the proxyIP is prohibited to prevent sniproxy going into a loop of requesting itself forever. Don't think it effects DoH though. Can't really tell without the full config. there's a chance that if your SOCKS5 is running on localhost, sniproxy outright rejects it to prevent such loops.

**Reply by holygrolli** [reply]
your last change in https://github.com/mosajjal/sniproxy/commit/caa674f0685841f9cab98d518fc1f1cbf933b5ee has broken my working basic udp dns setup with external socks5 proxy.
```
2023-05-15T20:53:00Z DBG app/acl/domain.go:151 > domain going through proxy: proxy.com acl=domain
2023-05-15T20:53:00Z WRN app/https.go:55 > read udp 172.18.0.2:48230->100.80.14.76:56193: i/o timeout service=https
```
while `100.80.14.76` being my socks5 on a different host. rolling back to my last self-compiled sniproxy it simply works:
```
2023-05-15T20:54:30Z DBG workspaces/sniproxy/acl/domain.go:151 > domain going through proxy: proxy.com acl=domain
2023-05-15T20:54:30Z INF workspaces/sniproxy/https.go:65 > establishing connection to targetip:443 from clientip:50064 with SNI proxy.com service=https
```

**Reply by mosajjal** [reply]
wondering if your SOCKS supports UDP or not. DNS used to be always direct and not proxied, now it is (experimentally). have you tested UDP connectivity on your SOCKS over Tailscale?

**Reply by holygrolli** [reply]
> wondering if your SOCKS supports UDP or not. DNS used to be always direct and not proxied, now it is (experimentally). have you tested UDP connectivity on your SOCKS over Tailscale?

I haven't tested it explicitly, but Dante is able to as I found out. This is in my logs:
```
Mai 15 20:11:24 ntflxproxy danted[22285]: May 15 20:11:24 (1684174284.006802) danted[22285]: info: pass(2): udp/udpassociate [: 0.0.0.0.0 100.80.14.76.56193
Mai 15 20:11:24 ntflxproxy danted[22285]: May 15 20:11:24 (1684174284.006802) danted[22285]: info: pass(2): udp/udpassociate [: 0.0.0.0.0 100.80.14.76.56193
Mai 15 20:11:24 ntflxproxy danted[23162]: May 15 20:11:24 (1684174284.014327) danted[23162]: notice: run_io(): 12 unexpected bytes over control connection from client 100.97.244.58.33292: \0\0\0\0\377\377\0\0\0\0\0\0
Mai 15 20:11:24 ntflxproxy danted[23162]: May 15 20:11:24 (1684174284.014327) danted[23162]: notice: run_io(): 12 unexpected bytes over control connection from client 100.97.244.58.33292: \0\0\0\0\377\377\0\0\0\0\0\0
```

can we make it configurable in the config like `upstream_dns_over_socks5: false`? I personally only need to transfer ProxyIP traffic through the socks5.

**Reply by mosajjal** [reply]
yeah we can do that. reckon that'll come in handy. ironically, I think if you have used a SOCKS that didn't support UDP it would've worked since I fall back to direct connection in those cases.
